spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindingListenerAnnotationBeanPostProcessor.java

```diff
@@ -0,0 +1,157 @@
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.SubscribableChannel;
+import org.springframework.messaging.core.DestinationResolver;
+import org.springframework.messaging.handler.annotation.SendTo;
+import org.springframework.messaging.handler.annotation.support.MessageHandlerMethodFactory;
+import org.springframework.messaging.handler.invocation.InvocableHandlerMethod;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author Marius Bogoevici
+ */
+public class BindingListenerAnnotationBeanPostProcessor implements BeanPostProcessor, SmartInitializingSingleton, ApplicationContextAware {
+
+	private final DestinationResolver<MessageChannel> binderAwareChannelResolver;
+
+	private final MessageHandlerMethodFactory messageHandlerMethodFactory;
+
+	private ConfigurableApplicationContext applicationContext;
+
+	public BindingListenerAnnotationBeanPostProcessor(DestinationResolver<MessageChannel> binderAwareChannelResolver, MessageHandlerMethodFactory messageHandlerMethodFactory) {
+		Assert.notNull(binderAwareChannelResolver, "Destination resolver cannot be null");
+		Assert.notNull(messageHandlerMethodFactory, "Message handler method factory cannot be null");
+		this.binderAwareChannelResolver = binderAwareChannelResolver;
+		this.messageHandlerMethodFactory = messageHandlerMethodFactory;
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = (ConfigurableApplicationContext) applicationContext;
+	}
+
+	@Override
+	public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+		return bean;
+	}
+
+	@Override
+	public Object postProcessAfterInitialization(final Object bean, String beanName) throws BeansException {
+		ReflectionUtils.doWithMethods(bean.getClass(), new ReflectionUtils.MethodCallback() {
+
+			@Override
+			public void doWith(final Method method) throws IllegalArgumentException, IllegalAccessException {
+				BindingListener bindingListener = AnnotationUtils.findAnnotation(method, BindingListener.class);
+				if (bindingListener != null) {
+					Assert.hasText(bindingListener.value(), "The binding name cannot be null");
+					final InvocableHandlerMethod invocableHandlerMethod = messageHandlerMethodFactory.createInvocableHandlerMethod(bean, method);
+					if (!StringUtils.hasText(bindingListener.value())) {
+						throw new BeanInitializationException("Only one of 'bindings' or 'value' can be specified");
+					}
+					SubscribableChannel channel = applicationContext.getBean(bindingListener.value(),
```

Comment:
There could be more than one listener wired to the same channel, maybe want to warn or have a 'strict' mode to control this case.